storm-core/src/jvm/backtype/storm/topology/WindowedBoltExecutor.java

```diff
@@ -0,0 +1,200 @@
+                    ((Number) stormConf.get(Config.TOPOLOGY_BOLTS_WINDOW_LENGTH_DURATION_MS)).intValue(),
+                    TimeUnit.MILLISECONDS);
+        }
+
+        if (stormConf.containsKey(Config.TOPOLOGY_BOLTS_SLIDING_INTERVAL_COUNT)) {
+            slidingIntervalCount = new Count(((Number) stormConf.get(Config.TOPOLOGY_BOLTS_SLIDING_INTERVAL_COUNT)).intValue());
+        } else if (stormConf.containsKey(Config.TOPOLOGY_BOLTS_SLIDING_INTERVAL_DURATION_MS)) {
+            slidingIntervalDuration = new Duration(((Number) stormConf.get(Config.TOPOLOGY_BOLTS_SLIDING_INTERVAL_DURATION_MS)).intValue(), TimeUnit.MILLISECONDS);
+        } else {
+            // default is a sliding window of count 1
+            slidingIntervalCount = new Count(1);
+        }
+        // validate
+        validate(stormConf, windowLengthCount, windowLengthDuration,
+                 slidingIntervalCount, slidingIntervalDuration);
+        if (windowLengthCount != null) {
+            manager.setWindowLength(windowLengthCount);
+        } else {
+            manager.setWindowLength(windowLengthDuration);
+        }
+        if (slidingIntervalCount != null) {
+            manager.setSlidingInterval(slidingIntervalCount);
+        } else {
+            manager.setSlidingInterval(slidingIntervalDuration);
+        }
+        return manager;
+    }
+
+    @Override
+    public void prepare(Map stormConf, TopologyContext context, OutputCollector collector) {
+        this.windowedOutputCollector = new WindowedOutputCollector(collector);
+        bolt.prepare(stormConf, context, windowedOutputCollector);
+        this.listener = newWindowLifecycleListener();
+        this.windowManager = initWindowManager(listener, stormConf);
+        LOG.info("Initialized window manager {} ", this.windowManager);
+    }
+
+    @Override
+    public void execute(Tuple input) {
+        windowManager.add(input);
+    }
+
+    @Override
+    public void cleanup() {
+        windowManager.shutdown();
+        bolt.cleanup();
+    }
+
+    @Override
+    public void declareOutputFields(OutputFieldsDeclarer declarer) {
+        bolt.declareOutputFields(declarer);
+    }
+
+    @Override
+    public Map<String, Object> getComponentConfiguration() {
+        return bolt.getComponentConfiguration();
+    }
+
+    private WindowLifecycleListener<Tuple> newWindowLifecycleListener() {
+        return new WindowLifecycleListener<Tuple>() {
+            @Override
+            public void onExpiry(List<Tuple> tuples) {
+                for (Tuple tuple : tuples) {
+                    windowedOutputCollector.ack(tuple);
+                }
+            }
+
+            @Override
+            public void onActivation(List<Tuple> tuples, List<Tuple> newTuples, List<Tuple> expiredTuples) {
+                windowedOutputCollector.setContext(tuples);
+                bolt.execute(new TupleWindowImpl(tuples, newTuples, expiredTuples));
+                newTuples.clear();
```

Comment:
Yes, here the caller should clear. I think I forgot to remove this after I refactored the code a bit. Will remove this.